sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUDFDynamicLoadSuite.scala

```diff
@@ -0,0 +1,254 @@
+      funcName: String,
+      jarUrl: String,
+      className: String,
+      // this function must call Hive UDF to load JAR
+      fnVerifyQuery: () => Unit): Unit = {
+    // force to use Spark classloader as other test (even in other test suites) may change the
+    // current thread's context classloader to jar classloader
+    Utils.withContextClassLoader(Utils.getSparkClassLoader) {
+      withUserDefinedFunction(funcName -> false) {
+        val sparkClassLoader = Thread.currentThread().getContextClassLoader
+
+        sql(s"CREATE FUNCTION $funcName AS '$className' USING JAR '$jarUrl'")
+
+        assert(Thread.currentThread().getContextClassLoader eq sparkClassLoader)
+
+        // JAR will be loaded at first usage, and it will change the current thread's
+        // context classloader to jar classloader in sharedState.
+        // See SessionState.addJar for details.
+        fnVerifyQuery()
+
+        assert(Thread.currentThread().getContextClassLoader ne sparkClassLoader)
+        assert(Thread.currentThread().getContextClassLoader eq
+          spark.sqlContext.sharedState.jarClassLoader)
+
+        // Roll back to the original classloader and run query again. Without this line, the test
+        // would pass, as thread's context classloader is changed to jar classloader. But thread
+        // context classloader can be changed from others as well which would fail the query; one
+        // example is spark-shell, which thread context classloader rolls back automatically. This
+        // mimics the behavior of spark-shell.
+        Thread.currentThread().setContextClassLoader(sparkClassLoader)
+
+        fnVerifyQuery()
+      }
+    }
+  }
+
+  private def testHiveUDFUsingJarWithChangingClassloaderWithCopyUDFExpression(
+      funcName: String,
+      jarUrl: String,
+      className: String,
+      // this function must call Hive UDF to load JAR
+      fnVerifyQuery: () => Unit,
+      fnCreateHiveUDFExpression: () => Expression): Unit = {
+    // force to use Spark classloader as other test (even in other test suites) may change the
+    // current thread's context classloader to jar classloader
+    Utils.withContextClassLoader(Utils.getSparkClassLoader) {
+      withUserDefinedFunction(funcName -> false) {
+        val sparkClassLoader = Thread.currentThread().getContextClassLoader
+
+        sql(s"CREATE FUNCTION $funcName AS '$className' USING JAR '$jarUrl'")
+
+        assert(Thread.currentThread().getContextClassLoader eq sparkClassLoader)
+
+        // JAR will be loaded at first usage, and it will change the current thread's
+        // context classloader to jar classloader in sharedState.
+        // See SessionState.addJar for details.
+        fnVerifyQuery()
+
+        assert(Thread.currentThread().getContextClassLoader ne sparkClassLoader)
+        assert(Thread.currentThread().getContextClassLoader eq
+          spark.sqlContext.sharedState.jarClassLoader)
+
+        val udfExpr = fnCreateHiveUDFExpression()
```

Comment:
I think this test should start here. The above test is the same as `testHiveUDFUsingJarWithChangingClassloader`.